### PR TITLE
Added docs/content/pytorch/concepts/using-functional-api/using-functional-api.md

### LINKED ISSUE #5864: [Concept Entry] PyTorch: Using Functional API
### Reason/inspiration (optional)

We would like a new entry on `Using Functional API` concept under PyTorch. The entry should go in a new file under `docs/content/pytorch/concepts/using-functional-api/using-functional-api.md`.       
  
The entry should include:      
  
- An introduction to the concept      
- A `Syntax` section that provides the syntax for the concept (if applicable)      
- An `Example` section that provides an example demonstrating the concept in use (if applicable)     

 
Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.|

### Entry Type (select all that apply)

New Concept Entry for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by ericgeiger1**
### Issue Solved

Closes #5864

### Type of Change

- Adding a new entry
- Editing an existing entry (fixing a typo, bug, issues, etc)
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

**Comment by ericgeiger1** [comment]
Waiting for approval on the issue, finished the project and just need to make the pull request and wait for review.

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+---
+
+**Functional APIs** in PyTorch provide a flexible and powerful way to define and manipulate neural networks. Unlike the `torch.nn.Module` class, which uses an object-oriented approach, functional APIs allow you to define models using functions. This can be particularly useful for creating complex models, experimenting with new architectures, or when you need more control over the forward pass.
+
+## Key Concepts
+
+### Functional Layers
+
+Functional layers are stateless and do not store parameters. Instead, parameters are passed explicitly. This can be useful for creating custom layers or when you need to reuse the same layer with different parameters.
+
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:23]: Please include a syntax section first before you move on to Example.

```suggestion
## Syntax

```pseudo
import torch
import torch.nn.functional as F

output = F.layer_name(input, *parameters, **kwargs)
```
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Example:
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:24]: Make this a heading.

```suggestion
## Example
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+x = torch.randn(10, 3)
+
+# Define weights and bias
+weight = torch.randn(5, 3, requires_grad=True)
+bias = torch.randn(5, requires_grad=True)
+
+# Apply linear transformation
+output = F.linear(x, weight, bias)
+```
+
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:40]: Add output for the above code like this:

```suggestion
The code will output a tensor of shape `(10, 5)` representing the result of the linear transformation. However, the actual values of the tensor will be random due to the use of `torch.randn`, which generates random values from a normal distribution. 

The output might look like this:

```shell
tensor([[-1.1570, -1.0890, -0.4154, -0.1795,  1.6989],
        [-0.5629, -0.3360, -0.3411, -0.2352,  1.0300],
        [-0.3185,  0.2398,  0.5389,  0.2491,  1.0749],
        ...
        [ 0.0665,  0.4579, -0.1494, -0.5361,  0.7465],
        [-0.5970,  0.3147,  0.1569,  0.1582,  0.5355],
        [-0.4481, -0.5795,  0.4445, -0.0623,  0.7024]])
```
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+### Custom Loss Functions
+
+Using functional APIs, you can easily define custom loss functions. This is useful when the predefined loss functions in PyTorch do not meet your requirements.
+
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:44]: Include syntax first and then move to example.

```suggestion
## Syntax

```pseudo
import torch

def custom_loss(output, target):
    # Define the loss calculation
    loss = some_loss_function(output, target)
    return loss
```
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+Example:
+```python
+def custom_loss(output, target):
+    loss = torch.mean((output - target) ** 2)
+    return loss
+
+# Example usage
+output = torch.randn(10, 5)
+target = torch.randn(10, 5)
+loss = custom_loss(output, target)
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:54]: Add a print statement to print the loss and also include the output here.

```suggestion
loss = custom_loss(output, target)
print("Custom MSE Loss:", loss)
```

The output might look something like this (the actual number will change due to randomness):

```shell
Custom MSE Loss: tensor(1.2135)
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+PyTorch provides a variety of activation functions in the `torch.nn.functional` module. These can be used to add non-linearity to your models.
+
+Example:
+```python
+x = torch.randn(10, 5)
+
+# Apply ReLU activation
+output = F.relu(x)
+```
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:67]: Similar comment, add syntax first, then add example, add output for example.

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+1. **Flexibility**: Functional APIs provide more control over the forward pass and allow for easy experimentation with different architectures.
+2. **Reusability**: Since functional layers are stateless, they can be reused with different parameters without any side effects.
+3. **Customizability**: Easily define custom layers, loss functions, and activation functions to suit your specific needs.
+
+## Conclusion
+
+Functional APIs in PyTorch offer a powerful and flexible way to define and manipulate neural networks. By understanding and utilizing these APIs, you can create more complex and customized models, experiment with new architectures, and have greater control over the forward pass.
+
+For more information, refer to the [PyTorch documentation](https://pytorch.org/docs/stable/nn.functional.html).
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:79]: We can skip conclusion section, docs concept entries are not that long that we would need a conclusion section.

```suggestion
```

**Review by PragatiVerma18 — CHANGES_REQUESTED** [review]
Hey @ericgeiger1, Thanks for contributing to Codecademy docs. I have some suggestions; please go through them and make the required changes. Let me know if you need any help or have any questions.

**Review by PragatiVerma18 — APPROVED** [review]
Hey @ericgeiger1, this is now ready for a second review! 🚀

**Review by ericgeiger1 — COMMENTED** [review]
This is my first pull request!

**Comment by Radhika-okhade** [comment]
@ericgeiger1 Please correct the file path. This is the correct path `docs/content/pytorch/concepts/using-functional-api/using-functional-api.md`

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
+---
+Title: 'Using Functional APIs'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:2]:
```suggestion
Title: 'Using Functional API'
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+Description: 'Functional APIs in PyTorch provide a flexible way to define and manipulate neural networks using functions rather than object-oriented classes.'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:3]:
```suggestion
Description: 'Functional API in PyTorch provide a flexible way to define and manipulate neural networks using functions rather than object-oriented classes.'
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+Subjects:
+  - 'Machine Learning'
+  - 'Data Science'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:6]:
```suggestion
  - 'Data Science'
  - 'Machine Learning' 
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
+---
+Title: 'Using Functional APIs'
+Description: 'Functional APIs in PyTorch provide a flexible way to define and manipulate neural networks using functions rather than object-oriented classes.'
+Subjects:
+  - 'Machine Learning'
+  - 'Data Science' 
+Tags:
+  - 'PyTorch'
+  - 'Deep Learning'
+  - 'Neural Networks'
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:10]:
```suggestion
  - 'Deep Learning'
  - 'Neural Networks'
  - 'PyTorch'
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+CatalogContent:
+  - 'learn-python'
+  - 'paths/data-science'
+---
+
+**Functional APIs** in PyTorch provide a flexible and powerful way to define and manipulate neural networks. Unlike the `torch.nn.Module` class, which uses an object-oriented approach, functional APIs allow you to define models using functions.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:16]:
```suggestion
**Functional API** in PyTorch provides a flexible and powerful way to define and manipulate neural networks. Unlike the `torch.nn.Module` class, which uses an object-oriented approach, functional API allows you to define models using functions. 
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+   - Bias: Optional bias terms can also be passed.
+- `**kwargs`: Represents additional keyword arguments that modify the behavior of the functional layer. Examples include:
+  - `Padding`: Used in convolutional layers to add padding.
+  - `Stride`: Specifies the step size for moving the kernel in convolution.
+  - `Dilation`: Controls the spacing between kernel elements in convolution.
+  - `Activation type`: Sometimes layers allow activation specification directly (e.g., `softmax` temperature).
+
+## Custom Loss Functions
+
+Using functional APIs, you can easily define custom loss functions. This is useful when the predefined loss functions in PyTorch do not meet your requirements.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:48]:
```suggestion
Using functional API, you can easily define custom loss functions. This is useful when the predefined loss functions in PyTorch do not meet your requirements.
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+- `output`: Represents the predicted values generated by the model. It is typically the output of the last layer of the model, such as logits, probabilities, or regression outputs.
+- `target`: Represents the true or ground-truth values corresponding to the predictions. It is used to compute the discrepancy between predictions and the actual values.
+- `loss`: This results from the loss function computation. It quantifies the error or difference between the predicted (`output`) and the actual (`target`) values. Common loss functions in PyTorch include:
+  - Mean Squared Error (MSE)
+  - Cross-Entropy Loss
+  - Binary Cross-Entropy Loss
+
+
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:68]: Only one line break is required. Remove the extra line break.

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+x = torch.randn(10, 3)
+
+# Define weights and bias
+weight = torch.randn(5, 3, requires_grad=True)
+bias = torch.randn(5, requires_grad=True)
+
+# Apply linear transformation
+output = F.linear(x, weight, bias)
+```
+The code will output a tensor of shape `(10, 5)` representing the result of the linear transformation. However, the actual values of the tensor will be random due to the use of `torch.randn`, which generates random values from a normal distribution.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:103]: Add a line break above this line.

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+output = F.relu(input)
+```
+
+- `input`: A tensor to which the ReLU activation will be applied. Negative values in the tensor will be replaced with zero.
+
+## Examples
+
+### Using Linear Transformation
+
+```python
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:89]:
```suggestion
```py
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+tensor([[0.0000, 0.2345, 1.4567, 0.0000, 0.9876],
+        [0.7654, 0.0000, 0.0000, 2.3456, 0.0000],
+        [0.0000, 0.0000, 1.1234, 0.5678, 0.0000],
+        ...,
+        [0.3456, 0.0000, 0.8765, 0.0000, 0.0000]])
+```
+
+> **Note**: Each run generates random values for `x`, so the exact output will vary.
+
+## Advantages of Using Functional APIs
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:166]:
```suggestion
## Advantages of Using Functional API
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+1. **Flexibility**: Functional APIs provide more control over the forward pass and allow for easy experimentation with different architectures.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:168]:
```suggestion
1. **Flexibility**: Functional API provide more control over the forward pass and allow for easy experimentation with different architectures.
```

File: content/pytorch/concepts/using-functional-apis/using-functional-api.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+CatalogContent:
+  - 'learn-python'
+  - 'paths/data-science'
+---
+
+**Functional APIs** in PyTorch provide a flexible and powerful way to define and manipulate neural networks. Unlike the `torch.nn.Module` class, which uses an object-oriented approach, functional APIs allow you to define models using functions. 
+
+This can be particularly useful for creating complex models, experimenting with new architectures, or needing more control over the forward pass.
+
+Functional APIs offer several advantages and unique features that set them apart. Below are essential concepts to understand how they work and how to use them effectively.
```
Comment (Radhika-okhade) [inline content/pytorch/concepts/using-functional-apis/using-functional-api.md:20]:
```suggestion
Functional API offer several advantages and unique features that set them apart. Below are essential concepts to understand how they work and how to use them effectively.
```

**Comment by Radhika-okhade** [comment]
Hey @ericgeiger1, Thanks for contributing to Codecademy docs. I have some suggestions; please go through them and make the required changes.

**Review by ericgeiger1 — COMMENTED** [review]
Thank You!

**Comment by Radhika-okhade** [comment]
Hey! @ericgeiger1, Please make the changes that I suggested.

**Review by Radhika-okhade — APPROVED** [review]

**Comment by Radhika-okhade** [comment]
Thank you for contributing to Codecademy docs @ericgeiger1. The entry looks good to be merged.
